### PR TITLE
Added code editor APIs for building responsive and switchable apps

### PR DESCRIPTION
**Pull request by mibrah42**
## What are we adding in this PR

This PR adds two new APIs on the code editor for creating responsive and switchable apps respectively. The two interfaces are shown below: 

**createMultiDeviceApp**

`createMultiDeviceApp` provides an API for building responsive templates. This could be used for building apps that dynamically resize to fit the window. The `createMultiDeviceApp` method takes in an object with two keys: `desktop` and `mobile`. Each key points to an app instance created by `createApp`. The following example will walk through building a multi-device app using this API.

```Javascript
var AppCreator = require(...);

/**
 * Create app instances.
 */
var desktop = AppCreator.createApp(...);
var mobile = AppCreator.createApp(...);

/*
 * ...
 * bind callbacks, add map layers, etc...
 * ...
*/

/**
 * Create a multiDevice app instance passing in the two apps created earlier.
 */
var responsiveApp = AppCreator.createMultiDeviceApp({desktop: desktop, mobile: mobile});

/**
 * Render app.
 */
responsiveApp.draw();
```

**Desktop**
Screenshot 2020-08-16 at 2 00 26 AM

**Mobile**

**createMultiSelectorApp**

`createMultiSelectorApp` provides an API for building switchable templates. This could be for switching between languages, themes, or app content. In the following example we are going to look at creating a multi-lingual app. The API takes in an optional object on construction. The object's keys are used as items on a ui.Select element. Each key points to an app instance created by either `createApp` or `createMultiDeviceApp`. This enables users to build switchable apps with responsive features.
 
If an initial object is not passed in, calling the set method can be used to add additional apps like so: `multiSelectorApp.add(key, app)`. Note that if the key already exists in the app, this method will overwrite it.
 
In this example, we're going to look at how to build a multi-lingual app in English, Spanish, and Arabic. This approach can also be applied to building multi-themed apps (i.e. Apps with light and dark modes), and apps with dynamic content (i.e. Each app performs a different type of analysis).

```Javascript
var AppCreator = require(...);

/**
 * Create app instances.
 */
var english = AppCreator.createApp(...);
var spanish = AppCreator.createApp(...);

/*
 * ...
 * bind callbacks, add map layers, etc...
 * ...
*/

/**
 * Create a multiSelectorApp instance passing in the two apps created earlier. The initial object is optional and apps can be added 
 * later on using the set method as shown in the following step.
 */
var multiSelectorApp = AppCreator.createMultiSelectorApp({English: english, Spanish: spanish});

/* 
 * Adding additional apps can be done like so.
 */
var arabic = AppCreator.createApp(...);

multiSelectorApp.set('Arabic', arabic);

/**
 * Render app.
 */
multiSelectorApp.draw();
```
**Demo**
capture (22)

### CODE REVIEW

File: app-creator/code-editor/deserializer.js [hunk 1]
```diff
@@ -440,17 +624,35 @@ function createApp(template) {
...
     app.root = app.deserializeUI(template);
-    app.templates.default = app.root;
   };
 
   /**
    * Draw UI to the screen by adding the root to ui.Root.
    */
-  app.draw = function () {
-    ui.root.widgets().reset([app.root]);
+  app.draw = function (container) {
```
Comment (michaelfdewitt) [inline app-creator/code-editor/deserializer.js:664]: Add a note about what the parameter is here.

File: app-creator/code-editor/deserializer.js [hunk 2]
```diff
@@ -208,10 +400,10 @@ function createApp(template) {
   app.createSelectElement = function (obj, style) {
     var uniqueAttributes = obj.uniqueAttributes;
 
-    var items = [''];
+    var items = ['Item 1'];
```
Comment (michaelfdewitt) [inline app-creator/code-editor/deserializer.js]: This seems like debug cruft.

File: app-creator/code-editor/deserializer.js [hunk 3]
```diff
@@ -2,32 +2,224 @@ var mapStyles = require('users/msibrahim/app-creator:map-styles');
...
+   * Callback triggered on window resize to conditionally render mobile and desktop templates.
    */
-  app.templates = {};
+  responsiveApp.resizeHandler = function (deviceInfo) {
+    /**
+     * We use the active property to know if this app is currently rendered.
+     * If it is not, we don't execute the onResize logic.
+     */
+    if (responsiveApp.active) {
+      if (!deviceInfo.is_desktop || deviceInfo.width < 900) {
```
Comment (michaelfdewitt) [inline app-creator/code-editor/deserializer.js]: I'd extract this to a constant.

File: app-creator/code-editor/deserializer.js [hunk 4]
```diff
@@ -2,32 +2,224 @@ var mapStyles = require('users/msibrahim/app-creator:map-styles');
 var sidemenu = require('users/msibrahim/app-creator:sidemenu');
 
 exports.createApp = createApp;
+exports.createResponsiveApp = createResponsiveApp;
```
Comment (michaelfdewitt) [inline app-creator/code-editor/deserializer.js]: Nit: the name here is a bit confusing, I think. I'm not sure that our non-dev-focused would be able to intuit what "responsive" means. Maybe something like "createCrossPlatformApp"?

File: app-creator/code-editor/deserializer.js [hunk 5]
```diff
@@ -2,32 +2,235 @@ var mapStyles = require('users/msibrahim/app-creator:map-styles');
...
 exports.createApp = createApp;
+exports.createMultiDeviceApp = createMultiDeviceApp;
+exports.createResponsiveApp = createResponsiveApp;
+exports.createMultiSelectorApp = createMultiSelectorApp;
+
+var MOBILE_WINDOW_SIZE = 900;
+
+/**
+ * Temporary placeholder to avoid breaking apps already created with this API.
+ */
```
Comment (mibrah42) [inline app-creator/code-editor/deserializer.js]: Is it too early for backward compatibility?

Comment (sufyanAbbasi) [reply app-creator/code-editor/deserializer.js]: I vote to mark it deprecated by throwing an error instead:
```
createResponsiveApp is deprecated, use createMultiDeviceApp instead.
```
But I'll let @michaelfdewitt have the final word here :)

File: app-creator/code-editor/deserializer.js (hunk 4 above)
Comment (michaelfdewitt) [reply app-creator/code-editor/deserializer.js]: This LGTM.

File: app-creator/code-editor/deserializer.js (hunk 5 above)
Comment (michaelfdewitt) [reply app-creator/code-editor/deserializer.js]: It's a good instinct, but I'd err on the side of removing this method entirely (while there are essentially zero callers). Do you have reason to think that anyone relies on the existence of this method today?

**Review by michaelfdewitt — APPROVED** [review]
Looks good to me!

File: app-creator/code-editor/deserializer.js (hunk 5 above)
Comment (mibrah42) [reply app-creator/code-editor/deserializer.js]: Just anyone who started playing around with the app creator after the project presentation.

Comment (michaelfdewitt) [reply app-creator/code-editor/deserializer.js]: Let's remove it then - reduced code complexity, particularly in the form of a smaller API surface, is well worth breaking some early dogfood users.

**Pull request merged**